File: snapshot/snapshot.py
import os
import time
import json
import requests
from web3 import Web3
from eth_account.messages import encode_defunct

# Load environment variables
eth_address = os.getenv('ETH_ADDRESS')
eth_private_key = os.getenv('ETH_PRIVATE_KEY')

if not eth_address or not eth_private_key:
    raise ValueError('Ethereum address or private key not provided in environment variables')

# Initialize Web3
w3 = Web3(Web3.HTTPProvider('https://rpc.ankr.com/eth_goerli'))  # Replace with your RPC endpoint

# Define proposal parameters
current_time = int(time.time())  # Current time in seconds
one_hour_in_seconds = 3600

proposal_params = {
    'space': 'testdao1.eth',
    'type': 'single-choice',  # define the voting system
    'title': 'Test proposal using Python',
    'body': 'This is the content of the proposal',
    'choices': ['Alice', 'Bob', 'Carol'],
    'start': current_time,
    'end': current_time + one_hour_in_seconds,  # End time is one hour later as an example
    'snapshot': w3.eth.block_number,  # Current block number as snapshot
    'network': '1',
    'plugins': json.dumps({}),
    'app': 'my-app'  # provide the name of your project using this Snapshot.js integration
}

# Sign the proposal
message = encode_defunct(text=json.dumps(proposal_params))
signed_message = w3.eth.account.sign_message(message, private_key=eth_private_key)

# Submit the proposal
url = 'https://testnet.hub.snapshot.org/api/message'
headers = {'Content-Type': 'application/json'}
data = {
    'address': eth_address,
    'msg': signed_message.messageHash.hex(),  # Use the hash of the original message
    'sig': signed_message.signature.hex(),
    'version': '2'
}
response = requests.post(url, headers=headers, data=json.dumps(data))

# Log the response details
if response.status_code == 200:
    print('Proposal submitted. Response:', response.json())
else:
    print('Error creating proposal:', response.text)

File: snapshot/snapshot.js
/**
 * This file contains the code for creating a proposal on Snapshot.
 * It uses the ethers.js library to interact with the blockchain.
 * The proposal is created using the Snapshot.js library.
 * The environment variables ETH_ADDRESS, ETH_PRIVATE_KEY, PRIMARY_RPC, and SECONDARY_RPC must be set in a .env file.
 * The ETH_ADDRESS is the address of the account that will create the proposal.
 * The ETH_PRIVATE_KEY is the private key of the ETH_ADDRESS account.
 * The PRIMARY_RPC is the primary RPC URL you wish to use for submitting proposals.
 * The SECONDARY_RPC is the secondary RPC URL you wish to use for submitting proposals.
 * wrapper.js is responsible for calling this function with the correct arguments.
 */
const { ethers, logger } = require('ethers');
const snapshot = require('@snapshot-labs/snapshot.js');
const dotenv = require('dotenv');

dotenv.config();

/**
 *
 * @param {string} title - The title of the proposal
 * @param {*} body - The body of the proposal
 * @param {*} choices - The choices for the proposal
 */
async function createProposal(title, body, choices) {
  const ethAddress = process.env.ETH_ADDRESS;
  const ethPrivateKey = process.env.ETH_PRIVATE_KEY;
  const primaryRpc = process.env.PRIMARY_RPC_URL;
  const secondaryRpc = process.env.SECONDARY_RPC_URL;
  const hub = process.env.SNAPSHOT_HUB;
  const snapshotSpace = process.env.SNAPSHOT_SPACE;
  const networkId = process.env.NETWORK_ID;

  const maxRetries = 3;
  const initialRetryDelay = 5000;

  if (!ethAddress || !ethPrivateKey) {
    throw new Error('Ethereum address or private key not provided in environment variables');
  }

  /**
   * Submits a proposal to Snapshot using the provided RPC endpoint
   * @param {string} providerRpc - RPC endpoint URL
   * @returns {Object} Object containing receipt and error
   */
  async function submitProposal(providerRpc) {
    const provider = new ethers.providers.JsonRpcProvider(providerRpc);
    const wallet = new ethers.Wallet(ethPrivateKey, provider);
    const client = new snapshot.Client712(hub);
    const currentTime = Math.floor(new Date().getTime() / 1000);
    const seventyTwoHours = 72 * 3600;

    try {
      const proposalParams = {
        space: snapshotSpace,
        type: 'weighted',
        title: title,
        body: body,
        choices: choices,
        start: currentTime,
        end: currentTime + seventyTwoHours,
        snapshot: await provider.getBlockNumber(),
        network: networkId,
        plugins: JSON.stringify({}),
        app: 'Gov'
      };

      logger.info(proposalParams)

      const receipt = await client.proposal(wallet, ethAddress, proposalParams);

      return { receipt, error: null };
    } catch (error) {
      console.error(`Error creating proposal with RPC ${providerRpc}:`, error);
      return { receipt: null, error };
    }
  }

  /**
   * Helper function to create a delay
   * @param {number} ms - Milliseconds to delay
   */
  function delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  let attempt = 0;
  let error = null;
  let receipt = null;
  let retryDelay = initialRetryDelay;
  let success = false;

  while (attempt < maxRetries && !success) {
    attempt++;
    console.log(`Attempt ${attempt} to submit proposal...`);

    const primaryResult = await submitProposal(primaryRpc);
    receipt = primaryResult.receipt;
    error = primaryResult.error;

    if (error) {
      console.log('Retrying with secondary RPC...');
      const secondaryResult = await submitProposal(secondaryRpc);
      receipt = secondaryResult.receipt;
      error = secondaryResult.error;
    }

    success = receipt !== null;

    if (error) {
      console.log(`Attempt ${attempt} failed.`);
      if (attempt < maxRetries) {
        console.log(`Waiting for ${retryDelay / 1000} seconds before retrying...`);
        await delay(retryDelay);
        retryDelay *= 2;
      }
    }
  }

  if (error) {
    console.error('Failed to submit proposal after multiple attempts with both RPC endpoints');
    return null
  } else {
    return receipt
  }
}

module.exports = createProposal;
